probably/pgcl/parser.py

```diff
@@ -313,30 +382,98 @@ def _parse_instrs(t: Tree) -> List[Instr]:
+        parameter = _parse_var(_child_tree(t, 1))
+        return OptimizationQuery(_parse_expr(_child_tree(t, 0)), parameter, opt_type)
+    elif t.data == "plot":
+        if len(t.children) == 3:
+            lit = _parse_literal(_child_tree(t, 2))
+            if isinstance(lit, BoolLitExpr):
+                raise SyntaxError("Plot instructions cannot handle boolean literals as arguments")
+            assert isinstance(t.children[2], Tree)
+            if t.children[2].data in ('real', 'infinity'):
+                assert isinstance(lit, RealLitExpr)
+                return PlotInstr(VarExpr(_parse_var(_child_tree(t, 0))),
+                                 VarExpr(_parse_var(_child_tree(t, 1))),
+                                 prob=lit)
+            else:
+                assert isinstance(lit, NatLitExpr)
+                return PlotInstr(VarExpr(_parse_var(_child_tree(t, 0))),
+                                 VarExpr(_parse_var(_child_tree(t, 1))),
+                                 term_count=lit)
+        elif len(t.children) == 2:
+            assert isinstance(t.children[1], Tree)
+            if t.children[1].data == 'real':
+                lit = _parse_literal(_child_tree(t, 1))
+                assert isinstance(lit, RealLitExpr)
+                return PlotInstr(VarExpr(_parse_var(_child_tree(t, 0))),
+                                 prob=lit)
+            elif t.children[1].data == 'nat':
+                lit = _parse_literal(_child_tree(t, 1))
+                assert isinstance(lit, NatLitExpr)
+                return PlotInstr(VarExpr(_parse_var(_child_tree(t, 0))),
+                                 term_count=lit)
+            elif t.children[1].data == 'infinity':
+                lit = _parse_literal(_child_tree(t, 1))
+                assert isinstance(lit, RealLitExpr)
+                return PlotInstr(VarExpr(_parse_var(_child_tree(t, 0))),
+                                 prob=lit)
+            elif t.children[1].data in ('true', 'false'):
+                raise SyntaxError("Plot instruction does not support boolean operators")
+            else:
+                return PlotInstr(VarExpr(_parse_var(_child_tree(t, 0))),
+                                 VarExpr(_parse_var(_child_tree(t, 1))),
+                                 )
+        else:
+            return PlotInstr(VarExpr(_parse_var(_child_tree(t, 0))))
+    else:
+        raise Exception(f'invalid AST: {t.data}')
+
+
+def _parse_program(config: ProgramConfig, t: Tree) -> Program:
     assert t.data == 'start'
     declarations = _parse_declarations(_child_tree(t, 0))
     instructions = _parse_instrs(_child_tree(t, 1))
-    return Program.from_parse(declarations, instructions)
+    instructions.extend(_parse_queries(_child_tree(t, 2)))
+    return Program.from_parse(config, declarations, instructions)
 
 
-def parse_pgcl(code: str) -> Program:
+def parse_pgcl(code: str, config: ProgramConfig = ProgramConfig()) -> Program:
```

Comment:
I think the `config` should be the first argument to this function.

Reply:
This would mean that we have to remove the default value though

Reply:
Oh right, then just leave as is.